### ISSUE TITLE
Explicit joins do not work with the InMemoryAdapter

### ISSUE THREAD

**Issue by Molokai** [ThatRendle/Simple.Data#142]
I have tried running joins with the database.User.Query().Join(database.Department).On().Select() syntax (where the columns to join on are specified, since there is no referential integrity with the InMemoryAdapter). The query returns data, but not joined data as far as I can tell.

All the selected fields from the Department table are null, and the number of rows returned are the number of rows in the User table. If I switch so that the first table is Department, then all the User fields are null and the number of rows is the same as in the Department table.

This was tested with 0.14.0.1

**Comment by ThatRendle** [comment]
Thanks for reporting this issue. I'll try to get it fixed in 0.15, which should be done in the next week.

Mark

**Comment by flq** [comment]
Moved to Issue #143

**Comment by Molokai** [comment]
Would natural joins work - if I used the not-so-pretty syntax to set up the relationships?

**Comment by ThatRendle** [comment]
Yes, and the call in your code under test could still use the explicit join; it gets ignored by InMemory at the moment.

**Comment by nbarnwell** [comment]
I'm working on GiveCRM and the following doesn't appear to return anything for "FacetValueId", "FacetId" or "FacetValueValue":

```
var query = db.Facets.All()
    .Select(
        db.Facets.Id,
        db.Facets.Type,
        db.Facets.Name,
        db.Facets.FacetValues.Id.As("FacetValueId"),
        db.Facets.FacetValues.FacetId,
        db.Facets.FacetValues.Value.As("FacetValueValue"))
    .OrderBy(db.Facets.Id);
```

Does this have any bearing on this issue or is it a different one?

**Comment by ThatRendle** [comment]
It's all part of the same issue, and will be fixed soon... honest...

**Comment by nbarnwell** [comment]
Cool. I've added the NUnit IgnoreAttribute to our failing integration tests on GiveCRM for now so when this fix is available I'll do an update and try to revive them.

**Comment by sagacity** [comment]
"where...in" constructions also don't seem to work with the InMemoryAdapter. Is this related?

E.g. `.FindAllByX(1)` filters correctly, but `.FindAllByX(new[] { 1 })` returns nothing.

**Comment by codereflection** [comment]
Hey Mark, do you have an idea on how to fix explicit joins in the InMemoryAdapter? I need it to work as well, and am willing to put in what time I can to get it fixed.  Just need to know where to start and how you think it should be done.

**Comment by BenMagyar** [comment]
So I ran into this on the .19 release, was this ever resolved in one of the 1.0+ pre-releases? Or is this something that was offset until the 2.0 release?
